RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+
+As an example, there’s a BC/FC breaking update on operator foo.
+
+Before:
+
+foo(Tensor self, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!)
+
+After:
+
+foo(Tensor self, **Tensor c**, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!)
+
+In schema, a Tensor argument, c is added. Note that it’s not added as a “tailing default argument”, so that BC/FC cannot be handled automatically.
+
+Accordingly, in the kernel of foo, the implementation is updated based on the new argument c. The pseudo code (it’s in Python format, but can be written in C++ as well) looks like:
+
+
+```
+def foo(Tensor self, Tensor c, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!):
+  # The original kernel implementation
+  ...
+  if not c.empty():
+    a.add_(c)
+```
+
+
+
+## Code changes (minimize the work of developer)
+
+If there is a BC/FC break with schema change in a PR, a lint error can be automatically generated, with instructions below to update the PR. 
+
+
+### Version bump
+
+Update version field in _operator_versions.yaml_.
+
+
+### BC updates
+
+The developer needs to implement a BC “upgrader” in Python. The upgrader code is put in a centralized python file, _operator_upgraders.py_. 
+
+
+
+<p id="gdcalert2" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: inline drawings not supported directly from Docs. You may want to copy the inline drawing to a standalone drawing and export by reference. See <a href="https://github.com/evbacher/gd2md-html/wiki/Google-Drawings-by-reference">Google Drawings by reference</a> for details. The img URL below is a placeholder. </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert3">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>
+
+
+![drawing](https://docs.google.com/drawings/d/12345/export/png)
+
+
+
+* Different from the upgraders defined[ here](https://github.com/pytorch/pytorch/blob/8a094e3270d2fbec6060099b7059898f4a1c104a/torch/csrc/jit/frontend/builtin_functions.cpp#L98), 
+* For some operator updates, it’s not possible to have a BC adapter. In such a case, the operator number could still help to check compatibility and to quickly detect the source of failure with meaningful error. 
+* For most of the operator changes, the upgrader code is not expected to be heavy. However, the performance overhead should be observed, especially for edge use cases.
+* If there are multiple upgraders (for example, v2 runtime loading v0 model), the upgraders must be chained: upgrader_0_1, then upgrader_1_2. 
+* If the breakage is hard (there is no upgrader), throw “version not supported” error and suggest to regenerate the model.
+
+
+### FC updates
+
+
+
+* No FC update is needed for
```

Comment:
This is probably the right section to deal with internal code refactors and how mobile tracing will be updated to handle them